pkg/plugin/manager.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/heptio/ark/pkg/cloudprovider"
+)
+
+type PluginKind string
+
+func (k PluginKind) String() string {
+	return string(k)
+}
+
+const (
+	PluginKindObjectStore PluginKind = "objectstore"
+	PluginKindBlockStore  PluginKind = "blockstore"
+)
+
+type pluginInfo struct {
+	kind PluginKind
+	name string
+}
+
+type Manager interface {
+	GetObjectStore(name string) (cloudprovider.ObjectStore, error)
+	GetBlockStore(name string) (cloudprovider.BlockStore, error)
+}
+
+type manager struct {
+	logger          hclog.Logger
+	clients         map[pluginInfo]*plugin.Client
+	internalPlugins map[pluginInfo]interface{}
+}
+
+func NewManager(logger logrus.FieldLogger, level logrus.Level) *manager {
+	return &manager{
+		logger:  (&logrusAdapter{impl: logger, level: level}).Named("plugin"),
+		clients: make(map[pluginInfo]*plugin.Client),
+		internalPlugins: map[pluginInfo]interface{}{
+			{kind: PluginKindObjectStore, name: "aws"}: struct{}{},
+			{kind: PluginKindBlockStore, name: "aws"}:  struct{}{},
+
+			{kind: PluginKindObjectStore, name: "gcp"}: struct{}{},
+			{kind: PluginKindBlockStore, name: "gcp"}:  struct{}{},
+
+			{kind: PluginKindObjectStore, name: "azure"}: struct{}{},
+			{kind: PluginKindBlockStore, name: "azure"}:  struct{}{},
+		},
+	}
+}
+
+func pluginDir() string {
+	pluginDir := os.Getenv("ARK_PLUGIN_DIR")
+	if pluginDir == "" {
+		pluginDir = "/plugins"
+	}
+
+	return pluginDir
+}
+
+func addPlugins(config plugin.ClientConfig, kinds ...PluginKind) plugin.ClientConfig {
+	for _, kind := range kinds {
+		if kind == PluginKindObjectStore {
+			config.Plugins[kind.String()] = &ObjectStorePlugin{}
+		} else if kind == PluginKindBlockStore {
+			config.Plugins[kind.String()] = &BlockStorePlugin{}
+		}
+	}
+
+	return config
+}
+
+func (m *manager) getPlugin(descriptor pluginInfo, logger hclog.Logger) (interface{}, error) {
+	client, found := m.clients[descriptor]
+	if !found {
+		var (
+			externalPath = filepath.Join(pluginDir(), fmt.Sprintf("ark-%s-%s", descriptor.kind, descriptor.name))
+			baseConfig   = plugin.ClientConfig{
+				HandshakeConfig:  Handshake,
+				AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+				Plugins:          make(map[string]plugin.Plugin),
+				Logger:           logger,
+			}
+		)
+
+		// First check to see if there's an external plugin for this kind and name. this
+		// is so users can override the built-in plugins if they want. If it doesn't exist,
+		// see if there's an internal one.
+		if _, err := os.Stat(externalPath); err == nil {
+			config := addPlugins(baseConfig, descriptor.kind)
+			config.Cmd = exec.Command(externalPath)
+
+			client = plugin.NewClient(&config)
+
+			m.clients[descriptor] = client
+		} else if _, found := m.internalPlugins[descriptor]; found {
+			config := addPlugins(baseConfig, PluginKindObjectStore, PluginKindBlockStore)
+			config.Cmd = exec.Command("/ark", "plugin", "cloudprovider", descriptor.name)
```

Comment:
So even if a plugin is internal, it's still invoked as a subprocess, rather than same-process?

Reply:
Yes, there's no way around that with the go-plugin package.